ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll

```diff
@@ -136,50 +140,65 @@ predicate returnStep(Node nodeFrom, Node nodeTo) {
  * literal string `"content"`, `nodeFrom` will be `y`, and `nodeTo` will be the
  * `Foo` object created on the first line of the function. This means we will
  * track the fact that `x.content` can have the type of `y` into the assignment
  * to `z` inside `bar`, even though this content write happens _after_ `bar` is
  * called.
  */
-predicate basicStoreStep(Node nodeFrom, Node nodeTo, string content) {
+predicate basicStoreStep(Node nodeFrom, Node nodeTo, TypeTrackerContentSet contents) {
+  postUpdateStoreStep(nodeFrom, nodeTo, contents)
+  or
+  exists(
+    SummarizedCallable callable, DataFlowPublic::CallNode call, SummaryComponent input,
+    SummaryComponent output
+  |
+    hasStoreSummary(callable, contents, input, output) and
+    call.asExpr().getExpr() = callable.getACallSimple() and
+    nodeFrom = evaluateSummaryComponentLocal(call, input) and
+    nodeTo = evaluateSummaryComponentLocal(call, output)
+  )
+}
+
+/**
+ * Holds if a store step `nodeFrom -> nodeTo` with `contents` exists, where the destination node
+ * is a post-update node that should be treated as a local source node.
+ */
+predicate postUpdateStoreStep(Node nodeFrom, Node nodeTo, TypeTrackerContentSet contents) {
   // TODO: support SetterMethodCall inside TuplePattern
   exists(ExprNodes::MethodCallCfgNode call |
-    content = getSetterCallAttributeName(call.getExpr()) and
+    contents
+        .isSingleton(DataFlowPublic::Content::getAttributeName(call.getExpr()
+                .(AST::SetterMethodCall)
+                .getTargetName())) and
     nodeTo.(DataFlowPublic::PostUpdateNode).getPreUpdateNode().asExpr() = call.getReceiver() and
     call.getExpr() instanceof AST::SetterMethodCall and
     call.getArgument(call.getNumberOfArguments() - 1) =
       nodeFrom.(DataFlowPublic::ExprNode).getExprNode()
   )
 }
 
-/**
- * Returns the name of the attribute being set by the setter method call, i.e.
- * the name of the setter method without the trailing `=`. In the following
- * example, the result is `"bar"`.
- *
- * ```rb
- * foo.bar = 1
- * ```
- */
-private string getSetterCallAttributeName(AST::SetterMethodCall call) {
-  // TODO: this should be exposed in `SetterMethodCall`
-  exists(string setterName |
-    setterName = call.getMethodName() and result = setterName.prefix(setterName.length() - 1)
-  )
-}
-
 /**
  * Holds if `nodeTo` is the result of accessing the `content` content of `nodeFrom`.
  */
-predicate basicLoadStep(Node nodeFrom, Node nodeTo, string content) {
+predicate basicLoadStep(Node nodeFrom, Node nodeTo, TypeTrackerContentSet contents) {
   exists(ExprNodes::MethodCallCfgNode call |
     call.getExpr().getNumberOfArguments() = 0 and
-    content = call.getExpr().getMethodName() and
+    contents.isSingleton(DataFlowPublic::Content::getAttributeName(call.getExpr().getMethodName())) and
     nodeFrom.asExpr() = call.getReceiver() and
     nodeTo.asExpr() = call
   )
+  or
+  exists(
```

Comment:
You should be able to use `summaryGetterStep` from `FlowSummaryImpl` instead.

Reply:
See answer above.

Reply:
Remember to also update this bit.